In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('Titanic.csv', na_values = ['?'])

<h1> 1. Inspekcja 20 pierwszych wierszy danych dotyczacych rejsu Titanica </h1>

In [3]:
df.head(20)

pclass  survived                                               name  \
0        1         1                      Allen, Miss. Elisabeth Walton   
1        1         1                     Allison, Master. Hudson Trevor   
2        1         0                       Allison, Miss. Helen Loraine   
3        1         0               Allison, Mr. Hudson Joshua Creighton   
4        1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5        1         1                                Anderson, Mr. Harry   
6        1         1                  Andrews, Miss. Kornelia Theodosia   
7        1         0                             Andrews, Mr. Thomas Jr   
8        1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9        1         0                            Artagaveytia, Mr. Ramon   
10       1         0                             Astor, Col. John Jacob   
11       1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12       1         1                      Aubart, Mme. Leontine Pauline   
13       1         1                       Barber, Miss. Ellen 'Nellie'   
14       1         1               Barkworth, Mr. Algernon Henry Wilson   
15       1         0                                Baumann, Mr. John D   
16       1         0                           Baxter, Mr. Quigg Edmond   
17       1         1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18       1         1                              Bazzani, Miss. Albina   
19       1         0                               Beattie, Mr. Thomson   

       sex      age  sibsp  parch    ticket      fare    cabin embarked boat  \
0   female  29.0000      0      0     24160  211.3375       B5        S    2   
1     male   0.9167      1      2    113781  151.5500  C22 C26        S   11   
2   female   2.0000      1      2    113781  151.5500  C22 C26        S  NaN   
3     male  30.0000      1      2    113781  151.5500  C22 C26        S  NaN   
4   female  25.0000      1      2    113781  151.5500  C22 C26        S  NaN   
5     male  48.0000      0      0     19952   26.5500      E12        S    3   
6   female  63.0000      1      0     13502   77.9583       D7        S   10   
7     male  39.0000      0      0    112050    0.0000      A36        S  NaN   
8   female  53.0000      2      0     11769   51.4792     C101        S    D   
9     male  71.0000      0      0  PC 17609   49.5042      NaN        C  NaN   
10    male  47.0000      1      0  PC 17757  227.5250  C62 C64        C  NaN   
11  female  18.0000      1      0  PC 17757  227.5250  C62 C64        C    4   
12  female  24.0000      0      0  PC 17477   69.3000      B35        C    9   
13  female  26.0000      0      0     19877   78.8500      NaN        S    6   
14    male  80.0000      0      0     27042   30.0000      A23        S    B   
15    male      NaN      0      0  PC 17318   25.9250      NaN        S  NaN   
16    male  24.0000      0      1  PC 17558  247.5208  B58 B60        C  NaN   
17  female  50.0000      0      1  PC 17558  247.5208  B58 B60        C    6   
18  female  32.0000      0      0     11813   76.2917      D15        C    8   
19    male  36.0000      0      0     13050   75.2417       C6        C    A   

     body                        home.dest  
0     NaN                     St Louis, MO  
1     NaN  Montreal, PQ / Chesterville, ON  
2     NaN  Montreal, PQ / Chesterville, ON  
3   135.0  Montreal, PQ / Chesterville, ON  
4     NaN  Montreal, PQ / Chesterville, ON  
5     NaN                     New York, NY  
6     NaN                       Hudson, NY  
7     NaN                      Belfast, NI  
8     NaN              Bayside, Queens, NY  
9    22.0              Montevideo, Uruguay  
10  124.0                     New York, NY  
11    NaN                     New York, NY  
12    NaN                    Paris, France  
13    NaN                              NaN  
14    NaN                    Hessle, Yorks  
15    NaN                  

Widzimy 20 pierwszych wierszy oraz kolumny: pclass, survived, name, sex, age, sibsp, parch, ticket, fare, cabin, embarked, boat, body i home.dest

<h1> 2. Sprawdzanie ilosci nulli w naszym datasecie. </h1>

In [4]:
df.isnull().sum()

pclass          0
survived        0
name            0
sex             0
age           263
sibsp           0
parch           0
ticket          0
fare            1
cabin        1014
embarked        2
boat          823
body         1188
home.dest     564
dtype: int64

In [5]:
df.isnull().mean()

pclass       0.000000
survived     0.000000
name         0.000000
sex          0.000000
age          0.200917
sibsp        0.000000
parch        0.000000
ticket       0.000000
fare         0.000764
cabin        0.774637
embarked     0.001528
boat         0.628724
body         0.907563
home.dest    0.430863
dtype: float64

In [6]:
null_count = df.isnull().sum().sum()
null_count

3855

In [7]:
percentage_of_nulls = null_count / df.size * 100
percentage_of_nulls

21.035687002073555

Braki danych wynosza 21.04% wszystkich danych - 3855 rekordow.

In [8]:
null_count = df.isnull().sum()
nulls = null_count/df.shape[0] * 100
nulls

pclass        0.000000
survived      0.000000
name          0.000000
sex           0.000000
age          20.091673
sibsp         0.000000
parch         0.000000
ticket        0.000000
fare          0.076394
cabin        77.463713
embarked      0.152788
boat         62.872422
body         90.756303
home.dest    43.086325
dtype: float64

<h1> 3. Rodzaj brakow danych wystepujacych w zbiorze danych (MCAR)/(MAR)/(MNAR) </h1>

Braki w danych (tzw. null'e), ktore wystepuja w naszym zbiorze danych wydaja sie nie byc przypadkowymi brakami - MNAR. Stanowia one ponad 21% calego zbioru, co nie pozwala na ich bezposrednie usuniecie. Odpowiedzia, dlaczego braki te moga nie byc przypadkowymi jest obserwacja kolumn np takich jak cabin oraz survived. Widzimy, ze dla ponad 87% przypadkow kiedy dany pasazer nie przezyl brakuje informacji o kabinie co potwierdza nasza hipoteze.

In [9]:
df['KolumnaNull'] = np.where(df['cabin'].isnull(), 1, 0)
df.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1   0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  KolumnaNull  
0                     St Louis, MO            0  
1  Montreal, PQ / Chesterville, ON            0  
2  Montreal, PQ / Chesterville, ON            0  
3  Montreal, PQ / Chesterville, ON            0  
4  Montreal, PQ / Chesterville, ON            0

In [10]:
df.groupby(['survived'])['KolumnaNull'].mean()

survived
0    0.873918
1    0.614000
Name: KolumnaNull, dtype: float64

Na podstawie powyzszego wyniku mozemy koncowo stwierdzic, ze faktycznie braki w danych sa typu MNAR. Mamy ponad 87% usrednionych 'matchow', kiedy w kolumnie kabin oraz survived sa zera - brak danych oraz smierc pasazera.

Jednolinijkowiec z dwoch powyzszych:

In [11]:
df['cabin'].isna().groupby(df['survived']).mean()

survived
0    0.873918
1    0.614000
Name: cabin, dtype: float64

In [12]:
df.isna().groupby(df['survived']).mean()

pclass  survived  name  sex       age  sibsp  parch  ticket  \
survived                                                                
0            0.0       0.0   0.0  0.0  0.234858    0.0    0.0     0.0   
1            0.0       0.0   0.0  0.0  0.146000    0.0    0.0     0.0   

              fare     cabin  embarked      boat      body  home.dest  \
survived                                                                
0         0.001236  0.873918     0.000  0.988875  0.850433   0.508035   
1         0.000000  0.614000     0.004  0.046000  1.000000   0.306000   

          KolumnaNull  
survived               
0                 0.0  
1                 0.0

<h1> 4. W jaki sposob postapic z brakujacymi wartosciami? </h1>

W przypadku naszego zbioru danych na pewno nie mozemy usunac wszystkich wierszy, w ktorych wystepuja braki w danych, poniewaz jak ustalilsmy braki te sa na poziomie ponad 21%, dodatkowo nie sa one spowodowane poprzez losowosc czy blad. Powinnismy podejsc indywiudalnie dla kazdej kolumny, w ktorej wystepuja braki: dla kolumn embarked i fare braki te sa szczatkowe i mozemy ja od razu usunac bez wiekszego wplywu na dalsze analizy, natomiast dla pozostalych mozemy postapic tak jak powyzej, czyli utworzyc kolumne binarna aby powiazac braki z pewna zaleznoscia (smierc pasazera), co mozemy wykorzystac dalej do budowy modelu ML.

In [13]:
print('Braki w kolumnie fare: ' + repr(df['fare'].isna().sum()) + ', w kolumnie embarked: ' + repr(df['embarked'].isna().sum()))

Braki w kolumnie fare: 1, w kolumnie embarked: 2


In [14]:
# df = df.dropna(subset = ['fare', 'embarked'])
# len(df) - 1306

<h1> Cwiczenie 3 </h1>

<h1> 1. Liczebnosc poszczegolnych etykiet dla zmiennych jakosciowych </h1>

In [15]:
print('Liczba etykiet zmiennej cabin: {}'.format(len(df.cabin.unique())))

Liczba etykiet zmiennej cabin: 187


<h1> 2. Liczba wszystkich pasazerow. </h1>

In [16]:
print('Liczba wszystkich pasazerow tragicznego rejsu: ' + repr(len(df.name.unique())))

Liczba wszystkich pasazerow tragicznego rejsu: 1307


<h1> 3. Podzial zbioru ze wzgledu na kardynalnosc (moc zbioru).

In [17]:
#x = [print('Liczba unikalnych etykiet kolumny ' + repr(i) + ' wynosi: ' + repr(len(df[i].unique()))) for i in df.columns]
for i in df.columns:
    print('Liczba unikalnych etykiet kolumny ' + repr(i) + ' wynosi: ' + repr(len(df[i].unique())))

Liczba unikalnych etykiet kolumny 'pclass' wynosi: 3
Liczba unikalnych etykiet kolumny 'survived' wynosi: 2
Liczba unikalnych etykiet kolumny 'name' wynosi: 1307
Liczba unikalnych etykiet kolumny 'sex' wynosi: 2
Liczba unikalnych etykiet kolumny 'age' wynosi: 99
Liczba unikalnych etykiet kolumny 'sibsp' wynosi: 7
Liczba unikalnych etykiet kolumny 'parch' wynosi: 8
Liczba unikalnych etykiet kolumny 'ticket' wynosi: 929
Liczba unikalnych etykiet kolumny 'fare' wynosi: 282
Liczba unikalnych etykiet kolumny 'cabin' wynosi: 187
Liczba unikalnych etykiet kolumny 'embarked' wynosi: 4
Liczba unikalnych etykiet kolumny 'boat' wynosi: 28
Liczba unikalnych etykiet kolumny 'body' wynosi: 122
Liczba unikalnych etykiet kolumny 'home.dest' wynosi: 370
Liczba unikalnych etykiet kolumny 'KolumnaNull' wynosi: 2


In [18]:
high_cardinality = []
low_cardinality = []

[high_cardinality.append(j) if len(df[j].unique()) > 100 else low_cardinality.append(j) for i, j in enumerate(df.columns)]
print('Kolumny o wysokiej kardynalnosci: ' + repr(high_cardinality) + '\nKolumny o niskiej kardynalnosci: ' + repr(low_cardinality))

Kolumny o wysokiej kardynalnosci: ['name', 'ticket', 'fare', 'cabin', 'body', 'home.dest']
Kolumny o niskiej kardynalnosci: ['pclass', 'survived', 'sex', 'age', 'sibsp', 'parch', 'embarked', 'boat', 'KolumnaNull']


Podzielilismy kolumny ze wzgledu na kardynalnosc (moc zbioru). Za wysoka kardynalnosc przyjalem ilosc etykiet powyzej 100 unikalnych. Otrzymalismy informacje, ze kolumnami o wysokiej kardynalnosci sa kolumny name, ticket, fare, cabin, body oraz home.dest. Niska kardynalnoscia odznaczaja sie pozostale.

Najwyzsza moca zbioru odznacza sie kolumna name - co jest zrozumiale, na 1309 pasazerow 1307 nazywalo sie inaczej od pozostalych. Wynik 187 dla kabin oznacza, ze na jedna kabine przypadalo kilku pasazerow - wedlug naszych statystyk 7 (1309/187). Wynik 929 dla kolumny ticket moze oznaczac, ze kilku pasazerow mialo bilet laczony, a home.dst na poziomie 370 oznacza, ze destynacje podrozy byly podobne dla wielu pasazerow.

In [19]:
z = np.array(df['cabin'].unique())
type(z)

numpy.ndarray

PandasDF zuzywaja wiecej pamieci, NumpyArray sa szybsze, dla danych tabelarycznych czesciej wykorzystuje sie PandasDataFrame, dla danych numerycznych raczej NumpyArray. PandasDF sa zbudowane z NumpyArray.

In [20]:
df['CabinReduced'] = df.cabin.astype(str).str[0]
df[['cabin', 'CabinReduced']].head(20)

cabin CabinReduced
0        B5            B
1   C22 C26            C
2   C22 C26            C
3   C22 C26            C
4   C22 C26            C
5       E12            E
6        D7            D
7       A36            A
8      C101            C
9       NaN            n
10  C62 C64            C
11  C62 C64            C
12      B35            B
13      NaN            n
14      A23            A
15      NaN            n
16  B58 B60            B
17  B58 B60            B
18      D15            D
19       C6            C

In [21]:
print('Liczba etykiet zmiennej CabinReduced: {}'.format(len(df.CabinReduced.unique())))

Liczba etykiet zmiennej CabinReduced: 9


In [22]:
print('Ilosc kolumn zredukowano o: '+ repr((1 - len(df.CabinReduced.unique()) / len(df.cabin.unique())) * 100) + '%')

Ilosc kolumn zredukowano o: 95.18716577540107%


Dokonuje redukcji tej zmiennej, poniewaz chcemy ulatwic zadanie dla modelu ML. Dzieki temu nasz model ML bedzie mial tylko 9 etykiet dla kabin, ktore wskaza na klase/sektor danej kabiny - analiza bedzie latwiejsza i szybsza. Za negatywne skutki mozemy uznac utrate niektorych informacji - dokladnosc polozenia danej kabiny ale trzeba odpowiedziec sobie na pytanie, czy taka informacja jest nam potrzebna do dalszej analizy.

<h1>Lab 3</h1>

Zadanie 1

In [24]:
df.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

       age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.0000      0      0   24160  211.3375       B5        S    2    NaN   
1   0.9167      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.0000      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.0000      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  KolumnaNull CabinReduced  
0                     St Louis, MO            0            B  
1  Montreal, PQ / Chesterville, ON            0            C  
2  Montreal, PQ / Chesterville, ON            0            C  
3  Montreal, PQ / Chesterville, ON            0            C  
4  Montreal, PQ / Chesterville, ON            0            C

Zadanie 2

Funkcja train_test_split to bardzo uzyteczne narzedzie z Scikit-learn, pozwala na bardzo szybki podzial danych wejsciowych na “wycinki”, czyli podzbiory na których będziemy trenowac podel, a nastepnie testowac, czy wyniki jego predykcji sa dla nas zadowalajace. Funkcja ma wiele uzytecznych parametrow, które możemy sobie modyfikowac, np. test_size, train_size - czyli w latwy i szybki sposob mozemy okreslic, jak duze chcemy miec nasze podzbiory. 

Zadanie 3

In [25]:
col_name = ['cabin', 'CabinReduced', 'sex']

Zadanie 4

In [34]:
df[['cabin']].values

array([['B5'],
       ['C22 C26'],
       ['C22 C26'],
       ...,
       [nan],
       [nan],
       [nan]], dtype=object)